src/Infrastructure.VS.UnitTests/VsUIServiceOperationTests.cs

```diff
@@ -0,0 +1,320 @@
+            var callback = new Mock<Func<IVsUIShell, string>>();
+
+            var testSubject = CreateTestSubject(serviceProvider.Object);
+
+            // 1. First call for IVsUIShell -> calls the service provider
+            _ = testSubject.Execute<SVsUIShell, IVsUIShell, string>(callback.Object);
+
+            serviceProvider.Verify(x => x.GetService(typeof(SVsUIShell)), Times.Once);
+            serviceProvider.Invocations.Should().HaveCount(1);
+            callback.Verify(x => x.Invoke(uiShellService), Times.Once);
+
+            // 2. Second call for the same service -> does not call the service provider
+            _ = testSubject.Execute<SVsUIShell, IVsUIShell, string>(callback.Object);
+
+            serviceProvider.Verify(x => x.GetService(typeof(SVsUIShell)), Times.Once);
+            serviceProvider.Invocations.Should().HaveCount(1);
+            callback.Verify(x => x.Invoke(uiShellService), Times.Exactly(2)); 
+        }
+
+        #endregion
+
+        #region Multiple calls, same different service
+
+        // NOTE - we're only testing this scenario for ExecuteAsync(Func)
+        // Adding tests for all of the other methods doesn't add enough value to justify the effort at this point.
+
+        [TestMethod]
+        public async Task ExecuteAsync_Func_MultipleCalls_DifferentServices_ServiceProviderIsCalledForEachService()
+        {
+            var service1 = Mock.Of<IVsUIShell>();
+            var service2 = Mock.Of<IVsMonitorSelection>();
+            var serviceProvider = new Mock<IServiceProvider>();
+            serviceProvider.Setup(x => x.GetService(typeof(SVsUIShell))).Returns(service1);
+            serviceProvider.Setup(x => x.GetService(typeof(SVsShellMonitorSelection))).Returns(service2);
+
+            var testSubject = CreateTestSubject(serviceProvider.Object);
+
+            // 1. First call for IVsUIShell -> calls the service provider
+            var callback1 = new Mock<Func<IVsUIShell, int>>();
+
+            await testSubject.ExecuteAsync<SVsUIShell, IVsUIShell, int>(callback1.Object);
+
+            serviceProvider.Verify(x => x.GetService(typeof(SVsUIShell)), Times.Once);
+            serviceProvider.Invocations.Should().HaveCount(1);
+            callback1.Verify(x => x.Invoke(service1), Times.Once);
+
+            // 2. Second call for a different service -> calls the service provider again
+            var callback2 = new Mock<Func<IVsMonitorSelection, int>>();
+            await testSubject.ExecuteAsync<SVsShellMonitorSelection, IVsMonitorSelection, int>(callback2.Object);
+
+            serviceProvider.Verify(x => x.GetService(typeof(SVsShellMonitorSelection)), Times.Once);
+            serviceProvider.Invocations.Should().HaveCount(2);
+            callback2.Verify(x => x.Invoke(service2), Times.Once);
+        }
+
+        #endregion
+
+        private static Mock<IServiceProvider> CreateConfiguredServiceProvider<S>(object serviceToReturn) 
```

Comment:
Question, couldn't you do the following so you could reuse this in the test method above here? 

```
private static Mock<IServiceProvider> CreateConfiguredServiceProvider<S>(param object[] servicesToReturn)
{
var serviceProvider = new Mock<IServiceProvider>();
foreach(var service in servicesToReturn)
{
            serviceProvider.Setup(x => x.GetService(typeof(S))).Returns(service );
}
return serviceProvider;
}
```

Reply:
Nice idea; I didn't think of that. It wouldn't quite work in that form since we want to set up different service types i.e. `<S>` changes each time. We could pass in an array of `(Type, Object)` tuples e.g.


```
        private static Mock<IServiceProvider> CreateConfiguredServiceProvider(params (Type serviceType, object serviceToReturn)[] serviceTuple)
        {
            var serviceProvider = new Mock<IServiceProvider>();
            foreach((var serviceType, var serviceToReturn) in serviceTuple)
            { 
                serviceProvider.Setup(x => x.GetService(serviceType)).Returns(serviceToReturn);
            }
            return serviceProvider;
        }
``` 


From the callers point of view, it works well if there are multiple services but it's a bit clunkier for a single service:
```
            // Multiple services
            var serviceProvider1 = CreateConfiguredServiceProvider(
                (typeof(SVsUIShell), service1),
                (typeof(SVsShellMonitorSelection), service2));

            // Single service
            var serviceProvider2 = CreateConfiguredServiceProvider(
                (typeof(SMyInterface), new object()));

            var serviceProvider3 = CreateConfiguredServiceProvider<SMyInterface>(new object());

```

I'll add it as an overload.